tools/print_tock_memory_usage.py

```diff
@@ -0,0 +1,422 @@
+    print("Embedded data (in flash): " + str(padding_text) + " bytes")
+    print()
+    function_groups = {}
+    # Embedded constants in code (e.g., after functions) aren't counted
+    # in the symbol's size, so detecting waste in code has too many false
+    # positives.
+    group_symbols(function_groups, kernel_functions, False, "Flash")
+    print_groups("Function groups (in flash)", function_groups)
+    print()
+
+def get_addr(symbol_entry):
+    """Helper function for sorting symbols by start address."""
+    return symbol_entry[1]
+
+def compute_padding(symbols):
+    """Calculate how much padding is in a list of symbols by comparing their
+       reporting size with the spacing with the next function and return
+       the total differences."""
+    symbols.sort(key=get_addr)
+    func_count = len(symbols)
+    diff = 0
+    for i in range(1, func_count):
+        (esymbol, eaddr, esize, _) = symbols[i - 1]
+        (_, laddr, _, _) = symbols[i]
+        total_size = laddr - eaddr
+        symbols[i - 1] = (esymbol, eaddr, esize, total_size)
+        if total_size != esize:
+            diff = diff + (total_size - esize)
+
+    return diff
+
+def parse_options(opts):
+    """Parse command line options."""
+    global symbol_depth, verbose, show_waste
+    valid = 'd:vs'
+    long_valid = ['depth=', 'verbose', 'show-waste']
+    optlist, leftover = getopt.getopt(opts, valid, long_valid)
+    for (opt, val) in optlist:
+        if opt == '-d' or opt == '--depth':
+            symbol_depth = int(val)
+        elif opt == '-v' or opt == '--verbose':
+            verbose = True
+        elif opt == '-s' or opt == '--show-waste':
+            show_waste = True
+        else:
+            usage("unrecognized option: " + opt)
+            return []
+
+    return leftover
+
+ # Script starts here ######################################
+arguments = sys.argv[1:]
+if len(arguments) < 1:
+    usage("no ELF specified")
+    sys.exit(-1)
+
+ # The ELF is always the last argument; pull it out, then parse
+ # the others.
+elf_name = ""
+options = arguments
+try:
+    remaining = parse_options(options)
+    if len(remaining) != 1:
+        usage("")
+        sys.exit(-1)
+    else:
+        elf_name = remaining[0]
+except getopt.GetoptError as err:
+    usage(str(err))
+    sys.exit(-1)
+
+header_lines = os.popen('arm-none-eabi-objdump -f ' + elf_name).readlines()
+
+print("Tock memory usage report for " + elf_name)
+arch = "UNKNOWN"
+
+for hline in header_lines:
+    # pylint: disable=anomalous-backslash-in-string
+    hmatch = re.search('file format (\S+)', hline)
+    if hmatch != None:
+        arch = hmatch.group(1)
+        if arch != 'elf32-littlearm':
+            usage(arch + " architecture not supported, only elf32-littlearm supportd")
+            sys.exit(-1)
+
+if arch == "UNKNOWN":
+    usage("could not detect architecture of ELF")
+    sys.exit(-1)
+
+objdump_lines = os.popen('arm-none-eabi-objdump -x ' + elf_name).readlines()
```

Comment:
why `arm-none-eabi-objdump`instead of `llvm-objdump`?

We are using `llvm-objdump` for tock so would make more sense to me and not have to introduce another dependency.

Reply:
Can we reuse the existing TOOLCHAIN/OBJDUMP variables we have defined in boards/Makefile.common? Passing the toolchain via environment variables is fairly standard and this way we could support both gcc/llvm toolchains.

Reply:
Reuse env variable OBJDUMP sounds good to me but I guess then `OBJDUMP` needs to be configured in the user's shell in order to work outside the `Makefile.common`.

Probably fine but I justed wanted to highlight it.

Reply:
Yeah, the tool doesn't include boards/Makefile.common. I'd rather not include shell variable dependencies.

Reply:
> We are using `llvm-objdump` for tock so would make more sense to me and not have to introduce another dependency.

Good point -- I will transition over. I used gcc because, well, it's the command I instinctively use.

Reply:
One question/issue here: do people have llvm-objdump in their PATH normally? I don't seem to have it, although I do of course have it within my rustup directories. Is there someone wrong with my configuration, or should the tool probe for/check for/find it?

Reply:
I transitioned to llvm-objdump.

Reply:
I have llvm-objdump in my path on Fedora